libflux/src/flux/semantic/fresh.rs

```diff
@@ -67,6 +67,22 @@ impl<T: Hash + Ord + Eq + Fresh, S> Fresh for HashMap<T, S> {
     }
 }
 
+impl<T: Fresh> Fresh for BTreeMap<String, T> {
+    fn fresh(self, f: &mut Fresher, sub: &mut HashMap<Tvar, Tvar>) -> Self {
+        self.into_iter()
+            .map(|(s, t)| (s, t.fresh(f, sub)))
+            .collect::<Self>()
+    }
+}
+
+impl<T: Hash + Ord + Eq + Fresh, S> Fresh for BTreeMap<T, S> {
+    fn fresh(self, f: &mut Fresher, sub: &mut HashMap<Tvar, Tvar>) -> Self {
+        self.into_iter()
+            .map(|(t, s)| (t.fresh(f, sub), s))
+            .collect::<Self>()
+    }
+}
+
 impl<T: Fresh> Fresh for Box<T> {
     fn fresh(self, f: &mut Fresher, sub: &mut HashMap<Tvar, Tvar>) -> Self {
         Box::new((*self).fresh(f, sub))
```

Comment:
Been a bit confused on the purpose of these, would be great to get a description for my understanding. It seems like many structures are endowed with fresher capabilities.

Reply:
This is mainly used for testing purposes. Tests involve type checking a flux program and asserting the identifiers in that program are of the correct type. Expected types are usually described using the language we have for type expressions:
```
forall [t0] (x: t0) -> t0
```

For example when type checking the following program:
```
...
f = (x) => x
...
```
type inference might assign the type `forall [t5] (x: t5) -> t5` to `f`. This type should compare as equal to `forall [t0] (x: t0) -> t0`. In order to do this we need to generate two fresh types that are structurally equivalent but have the same quantified type variables. This is what `Fresh` is used for.

Reply:
Note the types of the stdlib builtins are all specified using type expressions

https://github.com/influxdata/flux/blob/master/libflux/src/flux/semantic/builtins.rs.

Each one these type expressions is parsed and a new equivalent type with fresh type variables is generated before serialization. I believe this is the only place where `Fresh` is invoked for non-testing code.